ipfx/qc_features.py

```diff
@@ -117,3 +111,31 @@ def get_r_from_peak_pulse_response(v, i, t):
         r.append(r_instance)
 
     return np.mean(r)
+
+
+def get_square_pulse_idx(v):
+    """
+    Get up and down indices of the square pulse(s).
+    Skipping the very first pulse (test pulse)
+
+    Parameters
+    ----------
+    v: float
+        pulse trace
+
+    Returns
+    -------
+    up_idx, down_idx: list, list
+        up, down indices
+    """
+    dv = np.diff(v)
+
+    up_idx = np.flatnonzero(dv > 0)[1:] # skip the very first pulse (test pulse)
+    down_idx = np.flatnonzero(dv < 0)[1:]
+
+    assert len(up_idx) == len(down_idx), "Truncated square pulse"
```

Comment:
I wonder if it might be better to have [custom exceptions](https://docs.python.org/3/tutorial/errors.html#user-defined-exceptions) because in your tests, you expect an `AssertionError` but the source could come from either a `Truncated square pulse` or a `Negative pulse`.

Reply:
What would be the benefit of using custom exception here? Yes, the sources of error are different, but I do not need to handle them differently. If an error is raised then I will see the cause via the assertion statement.

Reply:
It matters for your tests, because you are doing `with pytest.raises(AssertionError):` if somewhere down the line, some regression is introduced that causes one assertion to fail but others to be fine, you could accidentally 'pass' a test that should fail.

Reply:
@njmei I matched the string representation of the exception. I think this will be good enough for such a test. Thanks for the suggestion.